From 360cb6550e6558606091ac8c64028ed9872ba6f9 Mon Sep 17 00:00:00 2001
From: Tanu Kaskinen <tanu.kaskinen@linux.intel.com>
Date: Wed, 7 Jan 2015 16:56:50 +0200
Subject: [PATCH 4/4] dynarray: Add PA_DYNARRAY_FOREACH

The PA_DYNARRAY_FOREACH macro requires that pa_dynarray_get() returns
NULL if the index is out of bounds.
---
 src/pulsecore/dynarray.c  | 4 +++-
 src/pulsecore/dynarray.h  | 5 +++++
 src/pulsecore/tokenizer.c | 3 ---
 3 files changed, 8 insertions(+), 4 deletions(-)

Index: pulseaudio/src/pulsecore/dynarray.c
===================================================================
--- pulseaudio.orig/src/pulsecore/dynarray.c	2015-08-27 17:22:56.068175130 +0200
+++ pulseaudio/src/pulsecore/dynarray.c	2015-08-27 17:22:56.060175027 +0200
@@ -72,7 +72,9 @@
 
 void *pa_dynarray_get(pa_dynarray *array, unsigned i) {
     pa_assert(array);
-    pa_assert(i < array->n_entries);
+
+    if (i >= array->n_entries)
+        return NULL;
 
     return array->data[i];
 }
Index: pulseaudio/src/pulsecore/dynarray.h
===================================================================
--- pulseaudio.orig/src/pulsecore/dynarray.h	2015-08-27 17:22:56.068175130 +0200
+++ pulseaudio/src/pulsecore/dynarray.h	2015-08-27 17:22:56.060175027 +0200
@@ -46,6 +46,8 @@
 void pa_dynarray_free(pa_dynarray *array);
 
 void pa_dynarray_append(pa_dynarray *array, void *p);
+
+/* Returns the element at index i, or NULL if i is out of bounds. */
 void *pa_dynarray_get(pa_dynarray *array, unsigned i);
 
 /* Returns the removed item, or NULL if the array is empty. */
@@ -53,4 +55,7 @@
 
 unsigned pa_dynarray_size(pa_dynarray *array);
 
+#define PA_DYNARRAY_FOREACH(elem, array, idx) \
+    for ((idx) = 0; ((elem) = pa_dynarray_get(array, idx)); (idx)++)
+
 #endif
Index: pulseaudio/src/pulsecore/tokenizer.c
===================================================================
--- pulseaudio.orig/src/pulsecore/tokenizer.c	2015-08-27 17:22:56.068175130 +0200
+++ pulseaudio/src/pulsecore/tokenizer.c	2015-08-27 17:22:56.060175027 +0200
@@ -78,8 +78,5 @@
 
     pa_assert(a);
 
-    if (i >= pa_dynarray_size(a))
-        return NULL;
-
     return pa_dynarray_get(a, i);
 }
